### ISSUE TITLE
C# mapscript broken? Crashes with: Attempted to read or write protected memory

### ISSUE THREAD

**Issue by BergWerkGIS** [mapserver/mapserver#5986]
## Problem

Some functions work, others do not and crash.
Simplest example I found:
* `mapscript.msGetVersionInt()`: ✅ 
* `mapscript.msGetVersion()`: 💥 

## OS and Versions

* Windows 10 Enterprise (1909, Build 18363.592)
* Visual Studio 2019 Enterprise 16.5.0 Preview 2.0
* dotnet core 2.2 and 3.1
* binaries from https://www.gisinternals.com/

Tried with all recent `MSVC 2017 x64` packages:
* `Stable Releases`
* `Stable Branches (daily)`
* `Development Versions (daily)`)

but had to go back to [`GDAL 2.4.0 MapServer 7.2.1`](http://download.gisinternals.com/sdk/downloads/release-1911-x64-gdal-2-4-0-mapserver-7-2-1.zip) to get W**M**S requests working.

Unfortunately with that old version W**F**S requests (only tried `GetCapabilities`) work exactly once and then also crash on `OWSDispatch()`. Maybe #5798 ?

## Test Case

I'm using `dotnet-script` (https://github.com/filipw/dotnet-script) for fast iteration but problem is the same with `dotnet run` commandline or debugging from within Visual Studio 2019.

* Download latest dev binaries (mine are from yesterday) http://download.gisinternals.com/sdk/downloads/release-1911-x64-gdal-mapserver.zip
* Extract to eg `.\dev-20200131\`
* Setup env vars: `.\dev-20200131\SDKShell.bat setenv`
* Change reference path to `mapscript_csharp.dll` in `test-umn.csx`
* Run test case: `dotnet script test-umn.csx`

**`test-umn.csx`**:
```csharp
#r "dev-20200131\bin\ms\csharp\mapscript_csharp.dll"
// ^^^^ change path to mapscript_csharp.dll here

using OSGeo.MapServer;

Console.WriteLine("----- before msGetVersionInt() -----");
Console.WriteLine($"UMN version  : {mapscript.msGetVersionInt()}");
Console.WriteLine("----- after msGetVersionInt() -----");

Console.WriteLine("----- before msGetVersion() -----");
Console.WriteLine($"UMN version full : {mapscript.msGetVersion()}");
Console.WriteLine("----- after msGetVersion() -----");
```

### Output

```
λ dotnet script test-umn.csx
----- before msGetVersionInt() -----
UMN version  : 70500
----- after msGetVersionInt() -----
----- before msGetVersion() -----
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at OSGeo.MapServer.mapscriptPINVOKE.msGetVersion()
   at OSGeo.MapServer.mapscriptPINVOKE.msGetVersion()
   at OSGeo.MapServer.mapscript.msGetVersion()
   at Submission#0+<<Initialize>>d__0.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.__Canon ByRef)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Int32, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[S
```

**Comment by BergWerkGIS** [comment]
Replacing `mapscript.dll` and `mapscript_csharp.dll` with the ones from the latest AppVeyor artifact (all other files still from gisinternals) seems to make my test case work:

```
λ dotnet script test-umn.csx
----- before msGetVersionInt() -----
UMN version  : 70500
----- after msGetVersionInt() -----
----- before msGetVersion() -----
UMN version full : MapServer version 7.5-dev OUTPUT=PNG OUTPUT=JPEG OUTPUT=KML SUPPORTS=PROJ SUPPORTS=AGG SUPPORTS=FREETYPE SUPPORTS=CAIRO SUPPORTS=SVG_SYMBOLS SUPPORTS=SVGCAIRO SUPPORTS=ICONV SUPPORTS=FRIBIDI SUPPORTS=WMS_SERVER SUPPORTS=WMS_CLIENT SUPPORTS=WFS_SERVER SUPPORTS=WFS_CLIENT SUPPORTS=WCS_SERVER SUPPORTS=SOS_SERVER SUPPORTS=FASTCGI SUPPORTS=THREADS SUPPORTS=GEOS SUPPORTS=PBF INPUT=JPEG INPUT=POSTGIS INPUT=OGR INPUT=GDAL INPUT=SHAPEFILE
----- after msGetVersion() -----
GDAL: In GDALDestroy - unloading GDAL shared library.
```

Hopefully I will find the time to check if that makes WMS/WFS work as well soon-ish.

**Comment by geographika** [comment]
@BergWerkGIS - which Appveyor artefact did you use? Was it the latest master?
I believe GISInternals uses SWIG 1.3 whereas Appveyor uses SWIG 3.0.12 which could explain the difference. 
If you get a chance - could you also test the artefacts from Appveyor #5983 which use SWIG 4.0.1?

**Comment by BergWerkGIS** [comment]
> which Appveyor artefact did you use? Was it the latest master?

Sorry for not being specify enough @geographika, just the latest in the build history:
https://ci.appveyor.com/project/MapServer/mapserver/builds/30440436

 > If you get a chance - could you also test the artefacts from Appveyor #5983 which use SWIG 4.0.1?

Will do, but probably won't get to it before next weekend.

----

Also found out more about WMS requests yesterday:
There seems to be a mishmash of proj.4 and proj.6 usage.
When pointing `PROJ_LIB` to proj.6 directory, `WMS GetCapabilities` fails because of missing `epsg` and `proj_def.dat`, which are proj.4.
So I copied them over. 
Then `GetCapabilities` works, but `GetMap` fails complaining about not finding `proj.db`, but I verified  it getting used by monitoring file access of the process.

Will do some more research and post my findings later this week.
Though I'm not sure to which repo, might be a build problem at gisinternals or maybe a mapserver problem.
Any suggestions?

**Comment by geographika** [comment]
GISInternals is using [1.3.39](https://github.com/gisinternals/buildsystem/blob/96939d0e83eb2aed4972378404a0cd62a4c1e520/Makefile#L558) by default (although you can get a build packages from GISInternals and set `SWIG_DIR` to point to any version when building locally). The last release was in 2009 so very likely .NET bindings would have problems (Python does).  

If the DLLs are fine with the Appveyor master build then there probably isn't a MapServer issue - we've dropped SWIG 1.x support. SWIG 2.x builds at the moment but not recommended. We might be recommending SWIG 4 in the next release of MapServer, but currently 3 and 4 both seem to be building without issues. 

I believe GISInternals may be planning a major refactoring of the build process so might be worth asking if the default version of SWIG can be set to 3.0.12 (or 4). 

Re the projection issues - maybe check if you get the same errors through MapServer CGI. If not then could it be a permissions issue?

**Comment by jmckenna** [comment]
@BergWerkGIS Thanks for providing your short test script.   I just tested this now with [MS4W 4.0.3](https://ms4w.com)     I already had Mono SDK installed locally, so here was my command and result:
```
"C:\Program Files (x86)\Mono\bin\csi" test-umn.csx

----- before msGetVersionInt() -----
UMN version  : 70600
----- after msGetVersionInt() -----
----- before msGetVersion() -----
UMN version full : MapServer version 7.6.0-dev (MS4W 4.0.3) OUTPUT=PNG OUTPUT=JPEG OUTPUT=KML SUPPORTS=PROJ SUPPORTS=AGG SUPPORTS=FREETYPE SUPPORTS=CAIRO SUPPORTS=SVG_SYMBOLS SUPPORTS=SVGCAIRO SUPPORTS=ICONV SUPPORTS=FRIBIDI SUPPORTS=WMS_SERVER SUPPORTS=WMS_CLIENT SUPPORTS=WFS_SERVER SUPPORTS=WFS_CLIENT SUPPORTS=WCS_SERVER SUPPORTS=SOS_SERVER SUPPORTS=FASTCGI SUPPORTS=THREADS SUPPORTS=GEOS SUPPORTS=POINT_Z_M SUPPORTS=PBF INPUT=JPEG INPUT=POSTGIS INPUT=OGR INPUT=GDAL INPUT=SHAPEFILE
----- after msGetVersion() -----
```
At least this shows that mapscript is not broken or causing crashes.

**Comment by jmckenna** [comment]
Here is the path that I used in your test-umn.csx script:
```
#r "D:\ms4w\msplugins\mapscript\csharp\mapscript_csharp.dll"
// ^^^^ change path to mapscript_csharp.dll here
```

**Comment by BergWerkGIS** [comment]
Thanks for testing @jmckenna.

> At least this shows that mapscript is not broken or causing crashes.

Agreed. 
As @geographika hinted, the errors I faced are mostly likely due to the older swig version used by gisinternals.

And the little `mapserver version` test case worked for me as well, after using a `mapscript_csharp.dll` from one of the latest mapserver AppVeyor builds.

-----

Got side tracked and haven't been able to provide the promised WMS/WFS test case yet.
But it's still on my list.

-----

Unfortunately with `ms4w@4.0.3` I get
```
Could not load file or assembly 'mapscript_csharp, Version=1.0.7331.21113, Culture=neutral, PublicKeyToken=80087771edd02926'.
```
 * downloaded the zip
* extracted to `C:\`
* called `C:\setenv.bat`
* adjusted the `#r` path

Btw, it would be super helpful to make `setenv.bat` more portable and not being tied to a fixed location (`\ms4w`).
Just in case you weren't aware, with `%~dp0` you get the directory where the bat is located.
Thinking about it, I suppose you already tried it and some of the tools absolutely need to know their location beforehand.


ms4w is a great package and has a lot of tools included to provide an all inclusive solution, but for my current use case (web app) it's a bit too heavy (400MB download and 1GB extracted).
I don't have access to the server and installing software there is a no go.
So I have to ship all necessary dlls as part of the, otherwise small and very lightweight, app.
For the gisinternals package I have a bat script that trims the ~110MB down to what I absolutely need: ~70MB.
Still a lot.

**Comment by jmckenna** [comment]
Here are my steps:
```
extract to C:/
cd /ms4w
setenv.bat
"C:\Program Files (x86)\Mono\bin\csi" test-umn.csx
```

I have just tried this now on several servers/machines.

**Comment by jmckenna** [comment]
Thanks for the complements on MS4W.  Indeed, quite a big product isn't it, much over 1GB, and comes with so many out-of-the-box OGC services running etc.   

Again, thanks for the complements.

**Comment by jmckenna** [comment]
@BergWerkGIS it would be great if you could file all of your enhancement requests in the MS4W issue tracker at https://ms4w.com/trac/report/1?sort=milestone&asc=1&page=1  Then all of the community benefits from your changes.

**Comment by jmckenna** [comment]
@BergWerkGIS  what is your exact #r path?  In your case, since you extracted to C it would be:
```
#r "C:\ms4w\msplugins\mapscript\csharp\mapscript_csharp.dll"
```

**Comment by jmckenna** [comment]
Or maybe it's an issue with your environment (mono versus dotnet-script).  Sorry for getting offtopic. Carry on :)

**Comment by BergWerkGIS** [comment]
> If you get a chance - could you also test the artefacts from Appveyor #5983 which use SWIG 4.0.1?

@geographika I know I'm late to the party, but these are my findings:

```
mapscript.msGetVersion()
```

. | gisinternals MSVC 2017 x64 daily dev Jan 31 2020 | gisinternals MSVC 2017 x64 daily dev Feb 19 2020
--- | ---- | ----
packaged `mapscript.dll/mapscript_csharp.dll` | 💥 | 💥 
`mapscript.dll/mapscript_csharp.dll` from [Pull request #5984 - msautotest: use Pytest](https://ci.appveyor.com/project/MapServer/mapserver/builds/30440436/job/c3i934j9gkjvb4dt/artifacts) | ✅ | 💥 
`mapscript.dll/mapscript_csharp.dll` from [Updates to build MapScript with SWIG 4.0.1 (#5983)](https://ci.appveyor.com/project/MapServer/mapserver/builds/30667048/job/qfck2fxh7s1hvcqi/artifacts) | ✅ | 💥 
`mapscript.dll/mapscript_csharp.dll` from [Merge pull request #5993 from geographika/mssql-wfs-tests](https://ci.appveyor.com/project/MapServer/mapserver/builds/30972142/job/dk03fchjdjlb4mgh/artifacts) | ✅ | 💥 
`mapscript.dll/mapscript_csharp.dll` from [ms4w_4.0.3.zip](https://ms4w.com/release/ms4w_4.0.3.zip) | 💥 | 💥



* **gisinternals Jan 31:** error is the same as before `System.AccessViolationException: Attempted to read or write protected memory.`
* **gisinternals Feb 19:** different error 
  ```
  System.EntryPointNotFoundException: Unable to find an entry point named 'CSharp_msGetVersionInt' in DLL 'mapscript'.
   at OSGeo.MapServer.mapscriptPINVOKE.msGetVersionInt()
   at OSGeo.MapServer.mapscript.msGetVersionInt()
  ```
* **ms4w@4.0.3:** I just noticed the dlls are 32 bit. I run everything 64 bit. Suppose that's why loading them into my process doesn't work.



-----

> For the gisinternals package I have a bat script that trims the ~110MB down to what I absolutely need

For these tests I used the unmodified gisinternals package (just replaced `mapscript/mapscript_csharp.dll`).

**Comment by BergWerkGIS** [comment]
Dang, `mapscript.msGetVersion()` works with current `mapserver@master` and `gisinternals dev` on AppVeyor.
Have to try that combination locally, or maybe also switch out `mapserver.dll` with my current setup.
But that's an exercise for another day.

https://ci.appveyor.com/project/BergWerkGIS/mapserver/builds/30977193#L636
![image](https://user-images.githubusercontent.com/4549888/75078918-3a54f600-5507-11ea-9017-6266012a2811.png)

**Comment by BergWerkGIS** [comment]
Replacing gisinternals' `mapserver.dll/mapscript.dll/mapscript_csharp.dll` with the ones from the Appveyor builds passes the minimal test case of `mapscript.msGetVersion()`.

However, the problem with proj.4 and proj.6 being needed for WMS requests still persists.
Will open a dedicated ticket once I can provide a trimmed down test case.

. | gisinternals MSVC 2017 x64 daily dev Jan 31 2020 | gisinternals MSVC 2017 x64 daily dev Feb 19 2020
--- | ---- | ----
`mapserver.dll/mapscript.dll/mapscript_csharp.dll` from [Pull request #5984 - msautotest: use Pytest](https://ci.appveyor.com/project/MapServer/mapserver/builds/30440436/job/c3i934j9gkjvb4dt/artifacts) | ✅ | ✅ 
`mapscript.dll/mapscript_csharp.dll` from [Updates to build MapScript with SWIG 4.0.1 (#5983)](https://ci.appveyor.com/project/MapServer/mapserver/builds/30667048/job/qfck2fxh7s1hvcqi/artifacts) | ✅ | ✅ 
`mapscript.dll/mapscript_csharp.dll` from [Merge pull request #5993 from geographika/mssql-wfs-tests](https://ci.appveyor.com/project/MapServer/mapserver/builds/30972142/job/dk03fchjdjlb4mgh/artifacts) | ✅ | ✅

**Comment by BergWerkGIS** [comment]
> However, the problem with proj.4 and proj.6 being needed for WMS requests still persists.
Will open a dedicated ticket once I can provide a trimmed down test case.

Won't be necessary.
After some more digging I found out that it is safe to ignore `GDAL/proj.6`'s error message about missing `proj.db` when accessing the data as mapserver uses the proj.4 it is linked against for reprojection and does not rely on GDAL to do so.

As I'm also using GDAL directly the single `PROJ_LIB` was a problem, but I solved it by merging `proj.4/share` and `proj.6/share` folders.
